### PR TITLE
GUACAMOLE-1881: Add parameter token containing the domain of LDAP/AD usernames.

### PR DESCRIPTION
**Pull request by mike-jumper**
This change is intended to supersede the changes proposed by poor @josnabattula via PR #931 that I have been very slow to review. It builds off the original commit (27bbd35a3d8d7bfb41ac50dfc145e1314121d289) by:

* Removing unnecessary use of `TokenName.canonicalize()` (there's no need to dynamically derive the token name of a static value).
* Decoupling the concept of extracting the domain from the concept of using that domain to create a token (`getDomainToken()` renamed and redocumented as `getUserDomain()`).
* Removing unnecessary recompilation of the domain extraction `Pattern` (there's no need to repeatedly recompile the regex each time a user logs in - it never changes).

### CODE REVIEW

**Review by josnabattula — APPROVED** [review]

**Comment by josnabattula** [comment]
@mike-jumper Can we get these changes into next release, its been months since I have raise pull request.

**Comment by josnabattula** [comment]
@mike-jumper  any plans to look into this pull request? it's been more than 6 months.

**Comment by mike-jumper** [comment]
@josnabattula Are you able to see Nick's comments above?

File: extensions/guacamole-auth-ldap/src/main/java/org/apache/guacamole/auth/ldap/AuthenticationProviderService.java [hunk 1]
```diff
@@ -63,7 +65,36 @@ public class AuthenticationProviderService {
...
      * The prefix that will be used when generating tokens.
      */
-    public static final String LDAP_ATTRIBUTE_TOKEN_PREFIX = "LDAP_";
+    public static final String LDAP_TOKEN_PREFIX = "LDAP_";
+
+    /**
+     * Regular expression that extracts the Windows / Active Directory domain
+     * from a username in either of the following formats: down-level logon
+     * ("DOMAIN\\username") or UPN ("username@domain"). If the username is in
+     * "DOMAIN\\username" format, the domain will be stored in the first
```
[the remark this answers is not in the document]
Comment (mike-jumper) [reply extensions/guacamole-auth-ldap/src/main/java/org/apache/guacamole/auth/ldap/AuthenticationProviderService.java:74]: @necouchman I ran into this recently, as well. Bizarrely, Java interprets backslash escapes in comments. Without the double backslash, that `\u` will be interpreted as an invalid Unicode escape and will not compile.

[the remark this answers is not in the document]
Comment (necouchman) [reply extensions/guacamole-auth-ldap/src/main/java/org/apache/guacamole/auth/ldap/AuthenticationProviderService.java:74]: Aha, okay, that's interesting. Thanks for clarifying.

**Comment by josnabattula** [comment]
> @josnabattula Are you able to see Nick's comments above?

I can see. But those comments are on the code you have included after my commit cheery-picked. I am thinking you are going to resolve it? Let me know if you want me to resolve those comments?

**Comment by mike-jumper** [comment]
Oof, right - that was silly of me. Sorry about that, @josnabattula.

File: extensions/guacamole-auth-ldap/src/main/java/org/apache/guacamole/auth/ldap/AuthenticationProviderService.java [hunk 2]
```diff
@@ -354,14 +425,19 @@ private Map<String, String> getAttributeTokens(ConnectedLDAPConfiguration config
...
             }
-
         }
         catch (LdapException e) {
             throw new GuacamoleServerException("Could not query LDAP user attributes.", e);
         }
 
+        // Extract the domain (ie: Windows / Active Directory domain) from the
+        // user's credentials
+        String domainName = getUserDomain(credentials);
```
[the remark this answers is not in the document]
Comment (mike-jumper) [reply extensions/guacamole-auth-ldap/src/main/java/org/apache/guacamole/auth/ldap/AuthenticationProviderService.java:437]: If assertions are enabled in the JVM (they are disabled by default) and there is a bug in the domain parsing that results in that assertion failing, the `AssertionError` will ultimately get caught by the error handling of the REST API and turn into an HTTP 503 response.

File: extensions/guacamole-auth-ldap/src/main/java/org/apache/guacamole/auth/ldap/AuthenticationProviderService.java [hunk 3]
```diff
@@ -308,25 +339,65 @@ public LDAPAuthenticatedUser authenticateUser(Credentials credentials)
     }
 
     /**
-     * Returns parameter tokens generated from LDAP attributes on the user
-     * currently bound under the given LDAP connection. The attributes to be
-     * converted into parameter tokens must be explicitly listed in
-     * guacamole.properties. If no attributes are specified or none are
-     * found on the LDAP user object, an empty map is returned.
+     * Returns the Windows / Active Directory domain included in the username
+     * of from the provided user credentials. If the username does not contain
```
[the remark this answers is not in the document]
Comment (mike-jumper) [reply extensions/guacamole-auth-ldap/src/main/java/org/apache/guacamole/auth/ldap/AuthenticationProviderService.java]: Fixed via rebase.

File: extensions/guacamole-auth-ldap/src/main/java/org/apache/guacamole/auth/ldap/AuthenticationProviderService.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (necouchman) [reply extensions/guacamole-auth-ldap/src/main/java/org/apache/guacamole/auth/ldap/AuthenticationProviderService.java:437]: Okay, seems safe to leave in there, the.

**Comment by josnabattula** [comment]
@necouchman Could you review again? All review comments are resolved now. thanks.

**Review by necouchman — APPROVED** [review]

**Pull request merged**